File: src/EfsAiHub.Core.Abstractions/Observability/RouterDecisionAnalytics.cs
namespace EfsAiHub.Core.Abstractions.Observability;

/// <summary>
/// Agregados de decisões de Router por projeto. Fonte: <c>workflow_event_audit</c>
/// filtrado em <c>EventType='node_completed'</c> + payload com
/// <c>agentType='Router'</c>. O output do Router já é JSON estruturado
/// (<c>{intent, confidence, reason, candidate_intents}</c>) — extraído
/// via cast jsonb dentro da query. Sem persistência nova; analytics opera
/// sobre o que o telemetria já loga.
///
/// Limitação conhecida: ~3% dos outputs vêm prefixados com "Assistant: ..."
/// (chain-of-thought de alguns provedores) e são filtrados fora da query.
/// Sub-amostragem em troca de robustez — analytics agregado tolera.
/// </summary>
public sealed class RouterDecisionOverview
{
    public required string ProjectId { get; init; }
    public DateTime PeriodFrom { get; init; }
    public DateTime PeriodTo { get; init; }

    /// <summary>Total de decisões parseáveis no período.</summary>
    public long TotalDecisions { get; init; }

    /// <summary>Quantidade de intents distintos vistos.</summary>
    public int DistinctIntents { get; init; }

    /// <summary>Confiança média (0..1) entre todas as decisões.</summary>
    public double AvgConfidence { get; init; }

    /// <summary>p50 da confiança (0..1).</summary>
    public double P50Confidence { get; init; }

    /// <summary>
    /// Decisões com confidence &lt; 0.5 — sinaliza ambiguidade ou modelo
    /// inseguro. Ratio = LowConfidence / TotalDecisions.
    /// </summary>
    public long LowConfidenceCount { get; init; }
    public double LowConfidenceRate { get; init; }

    /// <summary>
    /// Decisões com intent em <c>{out_of_scope, needs_clarification, loop_guard}</c>
    /// — fora de escopo, pediu esclarecimento, ou guardrail anti-loop bateu.
    /// Ratio = AmbiguityCount / TotalDecisions.
    /// </summary>
    public long AmbiguityCount { get; init; }
    public double AmbiguityRate { get; init; }

    public required IReadOnlyList<RouterIntentStats> Intents { get; init; }
}

/// <summary>Linha por intent decidido. Ordenada por count DESC.</summary>
public sealed class RouterIntentStats
{
    public required string Intent { get; init; }
    public long Count { get; init; }
    public double AvgConfidence { get; init; }
    public double P50Confidence { get; init; }
    public double P95Confidence { get; init; }
    public double MinConfidence { get; init; }
}

public sealed class RouterDecisionTimeseriesBucket
{
    public DateTime Bucket { get; init; }
    public long Total { get; init; }
    public double AvgConfidence { get; init; }
    public long LowConfidenceCount { get; init; }
    public long AmbiguityCount { get; init; }
}

File: analytics/src/api/routerDecisionAnalytics.ts
import { get } from './client'

// Espelha DTOs em
// src/EfsAiHub.Core.Abstractions/Observability/RouterDecisionAnalytics.cs.
// Endpoints em
// src/EfsAiHub.Host.Api/Controllers/RouterDecisionAnalyticsController.cs.
//
// Particularidade: ~3% dos outputs do Router têm prefixo "Assistant: " ou
// JSON malformado; a extração no backend descarta esses casos
// silenciosamente. Total reportado já é o "parseável" — sub-amostragem
// aceitável pra analytics agregado.

export interface RouterIntentStats {
  intent: string
  count: number
  avgConfidence: number
  p50Confidence: number
  p95Confidence: number
  minConfidence: number
}

export interface RouterDecisionOverview {
  projectId: string
  periodFrom: string
  periodTo: string
  totalDecisions: number
  distinctIntents: number
  avgConfidence: number
  p50Confidence: number
  lowConfidenceCount: number
  lowConfidenceRate: number
  ambiguityCount: number
  ambiguityRate: number
  intents: RouterIntentStats[]
}

export interface RouterDecisionTimeseriesBucket {
  bucket: string
  total: number
  avgConfidence: number
  lowConfidenceCount: number
  ambiguityCount: number
}

export type RouterGranularity = 'day' | 'hour'

function buildQs(projectId: string, from?: string, to?: string, extras?: Record<string, string>): string {
  const qs = new URLSearchParams()
  qs.set('projectId', projectId)
  if (from) qs.set('from', from)
  if (to) qs.set('to', to)
  if (extras) for (const [k, v] of Object.entries(extras)) qs.set(k, v)
  return `?${qs.toString()}`
}

export function getRouterDecisionSummary(
  projectId: string,
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<RouterDecisionOverview> {
  return get<RouterDecisionOverview>(`/analytics/router-decisions/summary${buildQs(projectId, from, to)}`, init)
}

export function getRouterDecisionTimeseries(
  projectId: string,
  groupBy: RouterGranularity = 'day',
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<RouterDecisionTimeseriesBucket[]> {
  return get<RouterDecisionTimeseriesBucket[]>(
    `/analytics/router-decisions/timeseries${buildQs(projectId, from, to, { groupBy })}`,
    init,
  )
}
